README.md
# Draw Graph

A dart package to draw line graphs in your flutter app.

<p align="center">
  <img src="https://user-images.githubusercontent.com/50942732/117781784-9caa2b80-b25e-11eb-9ede-de11542bd017.jpg" width="200" title="Graph">
  <img src="https://user-images.githubusercontent.com/50942732/117782154-fa3e7800-b25e-11eb-96da-4f6d53f1a08f.jpg" width="200" title="On Tapped">
</p>

## Usage

* Getting graph widget:
  ### LineGraph Widget
  #### Properties
  * features: [List](https://api.flutter.dev/flutter/dart-core/List-class.html)<feature>
    
    A list of features to be shown in the graph. Detailed description of Feature class below.
    (required)
  * size: [Size](https://api.flutter.dev/flutter/dart-ui/Size-class.html)
    
    This will determine the size of your graph. Height will be size.height - 50 in case when description is showed.
    (required)
  * labelX: [List](https://api.flutter.dev/flutter/dart-core/List-class.html)<[String](https://api.flutter.dev/flutter/dart-core/String-class.html)>
  
    A list of X-Axis Labels. This will determine and number of cells to distribute your data in and the width of your cells.
  * labelY: [List](https://api.flutter.dev/flutter/dart-core/List-class.html)<[String](https://api.flutter.dev/flutter/dart-core/String-class.html)>
  
    A list of Y-Axis Labels. The labels will be distributed on Y-Axis and determine height on cells.
  * fontFamily: [String](https://api.flutter.dev/flutter/dart-core/String-class.html)
  
    The fontFamily to use for labels and description of features.
  * graphColor: [Color](https://api.flutter.dev/flutter/dart-ui/Color-class.html)
  
    The color of your axis and labels.
  * showDescription: [bool](https://api.flutter.dev/flutter/dart-core/bool-class.html)
    
    Whether to show description at the end of graph.
  * graphOpacity: [double](https://api.flutter.dev/flutter/dart-core/double-class.html)
    
    The opacity of the area under graph.
  * verticalFeatureDirection: [bool](https://api.flutter.dev/flutter/dart-core/bool-class.html)
    
    Whether to scroll description vertically or not.
  * descriptionHeight: [double](https://api.flutter.dev/flutter/dart-core/double-class.html)
    
    If scrolling the description vertically, what should be the height of the description box.
  
  #### Constructor
  ```
  LineGraph({
    @required List<feature> features, 
    @required Size size,
    List<String> labelX: [],
    List<String> labelY: [],
    String fontFamily: 'sans serif',
    Color graphColor: Colors.grey,
    bool showDescription: false,
    double graphOpacity = 0.3,
    bool verticalFeatureDirection = false,
    double descriptionHeight = 80,
  });
  ```

  #### Example
  ```
  LineGraph({
    features: features,
    size: Size(320, 400),
    labelX: ['Day 1', 'Day 2', 'Day 3', 'Day 4', 'Day 5'],
    labelY: ['20%', '40%', '60%', '80%', '100%'],
    showDescription: true,
    graphColor: Colors.white30,
    graphOpacity: 0.2,
    verticalFeatureDirection: true,
    descriptionHeight: 130,
  })
  ```

* Creating a list of Features
  - import the Feature model
  ```import 'package:draw_graph/models/feature.dart';```
  
  ### Feature Model
  #### Properties
  * data: [List](https://api.flutter.dev/flutter/dart-core/List-class.html)<[double](https://api.flutter.dev/flutter/dart-core/double-class.html)>
  
    The Y-Axis values to be plotted. Should be in the range [0-1]. The length of the list should be equal to X-Axis labels.
    (required)
  * title: [String](https://api.flutter.dev/flutter/dart-core/String-class.html)
  
    The title displayed in Feature Description below the graph
  * color: [Color](https://api.flutter.dev/flutter/dart-ui/Color-class.html)
  
    The color of graph for this feature
    
  #### Constructor
  ```
  Feature({
    @required List<double> data,
    String title: '',
    Color color: Colors.black,
  });
  ```
  
  #### Example
  ```
  List<Feature> features = [
    Feature(
      title: "Drink Water",
      color: Colors.blue,
      data: [0.2, 0.8, 1, 0.7, 0.6],
    ),
    Feature(
      title: "Exercise",
      color: Colors.pink,
      data: [1, 0.8, 6, 0.7, 0.3, 8],
    ),
  ];
  ```

## Getting Started

See the `example` directory for a sample app using draw_graph.

CHANGELOG.md
## [0.0.1] - 2020-06-26

* First Version

## [0.0.2] - 2020-12-13

* Fixes issues

## [1.0.0] - 2021-05-11

* Adds feature - graph opacity
* Migrates to Null Safety

test/draw_graph_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:draw_graph/draw_graph.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/draw_graph.dart
library draw_graph;

import 'package:flutter/material.dart';

import 'package:draw_graph/models/feature.dart';
import 'package:draw_graph/widgets/lineGraph.dart';

class LineGraph extends StatefulWidget {
  final List<Feature> features;
  final Size size;
  final List<String>? labelX;
  final List<String>? labelY;
  final String? fontFamily;
  final Color graphColor;
  final bool showDescription;
  final double graphOpacity;
  final bool verticalFeatureDirection;
  final double descriptionHeight;

  LineGraph({
    required this.features,
    required this.size,
    this.labelX,
    this.labelY,
    this.fontFamily,
    this.graphColor = Colors.grey,
    this.showDescription = false,
    this.graphOpacity = 0.3,
    this.verticalFeatureDirection = false,
    this.descriptionHeight = 80,
  });

  @override
  _LineGraphState createState() => _LineGraphState();
}

class _LineGraphState extends State<LineGraph> {
  List<Feature>? features;
  bool tapped = false;

  @override
  Widget build(BuildContext context) {
    setState(() {
      if (!tapped) {
        features = widget.features;
      }
    });
    return Container(
      height: widget.size.height,
      width: widget.size.width,
      child: widget.showDescription
          ? widget.verticalFeatureDirection
              ? Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    getGraph(Size(widget.size.width,
                        widget.size.height - widget.descriptionHeight - 40)),
                    SizedBox(height: 40),
                    Container(
                      height: widget.descriptionHeight,
                      child: getFeautures(),
                    )
                  ],
                )
              : Column(
                  children: <Widget>[
                    getGraph(Size(widget.size.width, widget.size.height - 70)),
                    SizedBox(height: 40),
                    Container(
                      height: 28,
                      child: getFeautures(),
                    ),
                  ],
                )
          : getGraph(widget.size),
    );
  }

  Widget getGraph(Size size) {
    return CustomPaint(
      size: size,
      painter: LineGraphPainter(
        features: features,
        labelX: widget.labelX,
        labelY: widget.labelY,
        fontFamily: widget.fontFamily,
        graphColor: widget.graphColor,
        graphOpacity: widget.graphOpacity,
      ),
    );
  }

  Widget getFeautures() {
    List<Widget> featureDescriptions = [];
    for (int i = 0; i < features!.length; i++) {
      featureDescriptions.add(getDescription(features![i]));
    }

    return ListView(
      scrollDirection:
          widget.verticalFeatureDirection ? Axis.vertical : Axis.horizontal,
      children: featureDescriptions,
    );
  }

  Widget getDescription(Feature feature) {
    return Padding(
      padding: widget.verticalFeatureDirection
          ? EdgeInsets.only(left: 0, right: 0, bottom: 10, top: 10)
          : EdgeInsets.only(left: 20, right: 40, bottom: 0, top: 0),
      child: GestureDetector(
        child: Row(
          mainAxisSize: MainAxisSize.max,
          children: [
            Container(
              height: 24,
              width: 24,
              decoration: BoxDecoration(
                border: Border.all(color: feature.color, width: 1.5),
                color: feature.color.withOpacity(widget.graphOpacity),
              ),
            ),
            SizedBox(width: 20),
            Text(
              feature.title,
              style: TextStyle(
                fontFamily: widget.fontFamily,
              ),
            ),
          ],
        ),
        onTap: () {
          setState(() {
            if (!tapped) {
              features = [feature];
              tapped = true;
            } else {
              tapped = false;
            }
          });
        },
      ),
    );
  }
}


lib/widgets/lineGraph.dart
import 'package:flutter/material.dart';
import 'package:draw_graph/models/feature.dart';

class LineGraphPainter extends CustomPainter {
  final List<Feature>? features;
  final List<String>? labelX;
  final List<String>? labelY;
  final String? fontFamily;
  final Color graphColor;
  final double graphOpacity;

  LineGraphPainter({
    required this.features,
    required this.labelX,
    required this.labelY,
    required this.fontFamily,
    required this.graphColor,
    required this.graphOpacity,
  });

  @override
  void paint(Canvas canvas, Size size) {
    double _offsetX = 1;
    for (int i = 0; i < labelY!.length; i++) {
      if (labelY![i].length > _offsetX) {
        _offsetX = labelY![i].length.toDouble();
      }
    }

    _offsetX *= 7;
    _offsetX += 2 * size.width / 20;
    Size margin = Size(_offsetX, size.height / 8);
    Size graph = Size(
      size.width - 2 * margin.width,
      size.height - 2 * margin.height,
    );
    Size cell = Size(
      graph.width / (labelX!.length - 1),
      graph.height / labelY!.length,
    );

    drawAxis(canvas, graph, margin);
    drawLabelsY(canvas, size, margin, graph, cell);
    drawLabelsX(canvas, margin, graph, cell);

    for (int i = 0; i < features!.length; i++) {
      drawGraph(
        features![i],
        canvas,
        graph,
        cell,
        margin,
      );
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }

  void drawAxis(Canvas canvas, Size graph, Size margin) {
    Paint linePaint = Paint()
      ..color = graphColor
      ..strokeWidth = 1;

    Offset xEnd =
        Offset(graph.width + 2 * margin.width, graph.height + margin.height);
    Offset yStart = Offset(margin.width, 0);

    //X-Axis & Y-Axis
    canvas.drawLine(
        Offset(margin.width, graph.height + margin.height), xEnd, linePaint);
    canvas.drawLine(
        yStart, Offset(margin.width, graph.height + margin.height), linePaint);

    //Arrow heads
    canvas.drawLine(xEnd, Offset(xEnd.dx - 5, xEnd.dy - 5), linePaint);
    canvas.drawLine(xEnd, Offset(xEnd.dx - 5, xEnd.dy + 5), linePaint);
    canvas.drawLine(yStart, Offset(yStart.dx - 5, yStart.dy + 5), linePaint);
    canvas.drawLine(yStart, Offset(yStart.dx + 5, yStart.dy + 5), linePaint);
  }

  void drawLabelsY(
      Canvas canvas, Size size, Size margin, Size graph, Size cell) {
    for (int i = 0; i < labelY!.length; i++) {
      TextSpan span = new TextSpan(
        style: new TextStyle(
          color: graphColor,
          fontFamily: fontFamily,
        ),
        text: labelY![i],
      );
      TextPainter tp = new TextPainter(
        text: span,
        textAlign: TextAlign.left,
        textDirection: TextDirection.ltr,
      );
      tp.layout();
      tp.paint(
        canvas,
        new Offset(
          size.width / 20,
          margin.height + graph.height - 8 - (i + 1) * cell.height,
        ),
      );
    }
  }

  void drawLabelsX(Canvas canvas, Size margin, Size graph, Size cell) {
    for (int i = 0; i < labelX!.length; i++) {
      TextSpan span = new TextSpan(
        style: new TextStyle(
          color: graphColor,
          fontFamily: fontFamily,
        ),
        text: labelX![i],
      );
      TextPainter tp = new TextPainter(
          text: span,
          textAlign: TextAlign.left,
          textDirection: TextDirection.ltr);
      tp.layout();
      tp.paint(
        canvas,
        new Offset(
          margin.width + cell.width * i - 16,
          margin.height + graph.height + 10,
        ),
      );
    }
  }

  void drawGraph(
      Feature feature, Canvas canvas, Size graph, Size cell, Size margin) {
    Paint fillPaint = Paint()
      ..color = feature.color.withOpacity(graphOpacity)
      ..style = PaintingStyle.fill;
    Paint strokePaint = Paint()
      ..color = feature.color
      ..strokeWidth = 2
      ..style = PaintingStyle.stroke;

    Path path = Path();
    Path linePath = Path();
    path.moveTo(margin.width, graph.height + margin.height);
    path.lineTo(
      margin.width,
      margin.height + graph.height - feature.data[0] * graph.height,
    );
    linePath.moveTo(
      margin.width,
      margin.height + graph.height - feature.data[0] * graph.height,
    );
    int i = 0;
    for (i = 1; i < labelX!.length && i < feature.data.length; i++) {
      if (feature.data[i] > 1) {
        feature.data[i] = 1;
      }
      if (feature.data[i] < 0) {
        feature.data[i] = 0;
      }
      path.lineTo(
        margin.width + i * cell.width,
        margin.height + graph.height - feature.data[i] * graph.height,
      );
      linePath.lineTo(
        margin.width + i * cell.width,
        margin.height + graph.height - feature.data[i] * graph.height,
      );
    }
    // path.lineTo(
    //     margin.width + (feature.data.length - 1) * cell.width, margin.height);
    path.lineTo(
      margin.width + cell.width * (i - 1),
      margin.height + graph.height,
    );
    canvas.drawPath(path, fillPaint);
    canvas.drawPath(linePath, strokePaint);
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:draw_graph/draw_graph.dart';
import 'package:draw_graph/models/feature.dart';

final Color darkBlue = Color.fromARGB(255, 18, 32, 47);

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData.dark().copyWith(scaffoldBackgroundColor: darkBlue),
      home: Scaffold(
        appBar: AppBar(
          title: Text("DrawGraph Package"),
        ),
        body: MyScreen(),
      ),
    );
  }
}

class MyScreen extends StatelessWidget {
  final List<Feature> features = [
    Feature(
      title: "Drink Water",
      color: Colors.blue,
      data: [0.2, 0.8, 0.4, 0.7, 0.6],
    ),
    Feature(
      title: "Exercise",
      color: Colors.pink,
      data: [1, 0.8, 0.6, 0.7, 0.3],
    ),
    Feature(
      title: "Study",
      color: Colors.cyan,
      data: [0.5, 0.4, 0.85, 0.4, 0.7],
    ),
    Feature(
      title: "Water Plants",
      color: Colors.green,
      data: [0.6, 0.2, 0, 0.1, 1],
    ),
    Feature(
      title: "Grocery Shopping",
      color: Colors.amber,
      data: [0.25, 1, 0.3, 0.8, 0.6],
    ),
  ];

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.spaceAround,
      crossAxisAlignment: CrossAxisAlignment.center,
      children: <Widget>[
        Container(),
        Padding(
          padding: const EdgeInsets.symmetric(vertical: 64.0),
          child: Text(
            "Tasks Track",
            style: TextStyle(
              fontSize: 28,
              fontWeight: FontWeight.bold,
              letterSpacing: 2,
            ),
          ),
        ),
        LineGraph(
          features: features,
          size: Size(320, 400),
          labelX: ['Day 1', 'Day 2', 'Day 3', 'Day 4', 'Day 5'],
          labelY: ['20%', '40%', '60%', '80%', '100%'],
          showDescription: true,
          graphColor: Colors.white30,
          graphOpacity: 0.2,
          verticalFeatureDirection: true,
          descriptionHeight: 130,
        ),
        SizedBox(
          height: 50,
        )
      ],
    );
  }
}
